File: backend/src/controllers/brandsController.js
import { Brand } from "../models/Brand.js";
import { ok, created, badRequest, notFound } from "../utils/http.js";

/** Public: danh sách brands đang active (cho seller form, homepage) */
export async function listBrands(_req, res) {
  const docs = await Brand.find({ active: true })
    .sort({ name: 1 })
    .lean();
  const items = docs.map((d) => ({
    id: String(d._id),
    name: d.name,
    slug: d.slug || "",
  }));
  return ok(res, items);
}

/** Admin: danh sách tất cả brands */
export async function adminListBrands(_req, res) {
  const docs = await Brand.find().sort({ name: 1 }).lean();
  const items = docs.map((d) => ({
    id: String(d._id),
    name: d.name,
    slug: d.slug || "",
    active: d.active !== false,
  }));
  return ok(res, items);
}

/** Admin: tạo brand mới */
export async function adminCreateBrand(req, res) {
  const { name, slug } = req.body || {};
  const trimmed = typeof name === "string" ? name.trim() : "";
  if (!trimmed) return badRequest(res, "Brand name is required");

  const existing = await Brand.findOne({ name: { $regex: new RegExp(`^${trimmed}$`, "i") } });
  if (existing) return badRequest(res, "Brand already exists");

  const doc = await Brand.create({
    name: trimmed,
    slug: typeof slug === "string" ? slug.trim() : trimmed.toLowerCase().replace(/\s+/g, "-"),
    active: true,
  });
  const out = doc.toJSON ? doc.toJSON() : doc;
  return created(res, out);
}

/** Admin: cập nhật brand */
export async function adminUpdateBrand(req, res) {
  const { id } = req.params;
  const { name, slug, active } = req.body || {};

  const doc = await Brand.findById(id);
  if (!doc) return notFound(res, "Brand not found");

  if (typeof name === "string" && name.trim()) {
    const trimmed = name.trim();
    const existing = await Brand.findOne({
      name: { $regex: new RegExp(`^${trimmed}$`, "i") },
      _id: { $ne: id },
    });
    if (existing) return badRequest(res, "Brand name already exists");
    doc.name = trimmed;
  }
  if (typeof slug === "string") doc.slug = slug.trim();
  if (typeof active === "boolean") doc.active = active;

  await doc.save();
  const out = doc.toJSON ? doc.toJSON() : doc;
  return ok(res, out);
}

/** Admin: xóa brand */
export async function adminDeleteBrand(req, res) {
  const { id } = req.params;
  const doc = await Brand.findByIdAndDelete(id);
  if (!doc) return notFound(res, "Brand not found");
  return ok(res, { deleted: true, id });
}

File: backend/src/main/java/com/biketrading/backend/controller/BrandsController.java
package com.biketrading.backend.controller;

import com.biketrading.backend.entity.Brand;
import com.biketrading.backend.repository.BrandRepository;
import com.biketrading.backend.util.ApiResponse;
import com.biketrading.backend.util.MapperUtil;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@RestController
public class BrandsController {
    private final BrandRepository brandRepository;

    public BrandsController(BrandRepository brandRepository) {
        this.brandRepository = brandRepository;
    }

    @GetMapping("/api/brands")
    public ResponseEntity<?> listActive() {
        List<?> items = brandRepository.findByActiveTrueOrderByNameAsc().stream().map(MapperUtil::brandDto).collect(Collectors.toList());
        return ApiResponse.ok(items);
    }

    @GetMapping("/api/admin/brands")
    public ResponseEntity<?> adminList() {
        List<?> items = brandRepository.findAllByOrderByNameAsc().stream().map(MapperUtil::brandDto).collect(Collectors.toList());
        return ApiResponse.ok(items);
    }

    @PostMapping("/api/admin/brands")
    public ResponseEntity<?> create(@RequestBody Map<String, Object> body) {
        String name = String.valueOf(body.getOrDefault("name", "")).trim();
        if (name.isBlank()) return ApiResponse.error(HttpStatus.BAD_REQUEST, "Brand name is required");
        if (brandRepository.findByNameIgnoreCase(name).isPresent()) return ApiResponse.error(HttpStatus.BAD_REQUEST, "Brand already exists");
        Brand b = new Brand();
        b.setName(name);
        String slug = String.valueOf(body.getOrDefault("slug", name.toLowerCase().replace(" ", "-"))).trim();
        b.setSlug(slug.isBlank() ? name.toLowerCase().replace(" ", "-") : slug);
        b.setActive(true);
        brandRepository.save(b);
        return ApiResponse.created(MapperUtil.brandDto(b));
    }

    @PutMapping("/api/admin/brands/{id}")
    public ResponseEntity<?> update(@PathVariable Long id, @RequestBody Map<String, Object> body) {
        Brand b = brandRepository.findById(id).orElse(null);
        if (b == null) return ApiResponse.error(HttpStatus.NOT_FOUND, "Brand not found");
        if (body.containsKey("name")) {
            String name = String.valueOf(body.get("name")).trim();
            if (!name.isBlank()) b.setName(name);
        }
        if (body.containsKey("slug")) b.setSlug(String.valueOf(body.get("slug")).trim());
        if (body.containsKey("active")) b.setActive(Boolean.parseBoolean(String.valueOf(body.get("active"))));
        brandRepository.save(b);
        return ApiResponse.ok(MapperUtil.brandDto(b));
    }

    @DeleteMapping("/api/admin/brands/{id}")
    public ResponseEntity<?> delete(@PathVariable Long id) {
        Brand b = brandRepository.findById(id).orElse(null);
        if (b == null) return ApiResponse.error(HttpStatus.NOT_FOUND, "Brand not found");
        brandRepository.delete(b);
        return ApiResponse.ok(Map.of("deleted", true, "id", String.valueOf(id)));
    }
}
